[WIP] Q1 dimension-dependent real-to-unit transforms

drwells opened the pull request:
This is a work-in-progress because it causes a dozen tests (which test which cell an edge point or corner is in) to fail and I am not sure if breaking that behavior is significant. The failing tests are:

```
1358 - bits/find_cell_4.debug (Failed)
1405 - bits/find_cell_alt_4.release (Failed)
1698 - bits/find_cell_4.release (Failed)
1899 - bits/find_cell_alt_4.debug (Failed)
2003 - bits/find_cell_5.release (Failed)
2053 - bits/find_cell_1.debug (Failed)
2054 - bits/find_cell_1.release (Failed)
2208 - bits/find_cell_alt_1.debug (Failed)
2209 - bits/find_cell_alt_1.release (Failed)
2458 - bits/find_cell_alt_5.debug (Failed)
2663 - bits/find_cell_alt_5.release (Failed)
2759 - bits/find_cell_5.debug (Failed)
```

The 2D transform with the exact formula is about 45x faster and the 3D transform is about 8x faster. The accuracy is about the same (I think the new 3D solver is a bit more accurate). My transforms perform badly when the point is not in the cell (I switch back to the original algorithm in those cases) so this does not quite solve #314, #866, or #872.

kronbichler commented on source/fe/mapping_q1_eulerian.cc:
```diff
@@ -36,66 +37,66 @@ MappingQ1Eulerian (const EulerVectorType  &euler_transform_vectors,
+  for (unsigned int i = 0; i < GeometryInfo<dim>::vertices_per_cell; ++i)
+    {
+      a[i] = vertices[i];
     }
```
This is now the same function as in MappingQ1, isn't it? In that case, remove it.

kronbichler commented:
I guess you have checked that allocating the memory on the stack with std::array rather than on the heap with std::vector makes sense performance-wise for get_vertices(). Even in 3D where we get 24 double values, getting a memory region from malloc may take ten times as much time. Further, passing a large object around by value seems inefficient at frist but this should be the classical case of return value optimization that avoids temporaries, so I think the code should show good performance.

kronbichler commented on source/fe/mapping_q1.cc:
```diff
@@ -92,6 +93,246 @@ namespace internal
+            {
+              break;
+            }
+
```
This is a classical Newton algorithm for finding the point location. Please add some comment at the top of this loop.

bangerth commented on source/fe/mapping_q1.cc:
```diff
@@ -92,6 +93,246 @@ namespace internal
+        {
+          const double x8 = eta*x3;
+          const double x9 = -eta*x2;
+          const double x10 = eta - 1;
```
Here and below, is this code automatically generated? If so, can you also add (as a comment) the original program from which it was created?

Otherwise, all of this is so much easier to write and understand if it were stated in terms of the functionality provided by the `Tensor<1,dim>` and `Tensor<2,dim>` classes which allow you to write all of this in terms of vectors and matrices, their inverses, their determinants, etc.

drwells commented:
@kronbichler Yes, I was a bit surprised how much this helped: the 2D code is merely 20x faster without using `std::array`.The 3D code is only 7x faster. I only profiled the Q1 mapping. I would be surprised if there is a case where, as long as the compiler does RVO, the performance is worse than using `std::vector<Point<dim>>`.

 It also avoids writing much worse template code because otherwise `dim` only shows up in the output value and that cannot be used for type deduction.

@bangerth I know it looks like FORTRAN-77. I generated some of it with `sage` but a lot of it was by hand. I will see if I can write it to not be so awful with the tensors.

drwells commented:
The 3D version I wrote fails about 70% of the time when the cells are not rectangular prisms (where the cartesian mapping would be faster anyway) so I removed it. Otherwise I addressed the comments, but
```
2055 - bits/find_cell_1.debug (Failed)
2056 - bits/find_cell_1.release (Failed)
2210 - bits/find_cell_alt_1.debug (Failed)
2211 - bits/find_cell_alt_1.release (Failed)
```
still fail.

tamiko commented:
@drwells Can you please rebase?

@kronbichler  This pull request is good to be merged. Any final remarks?

kronbichler commented:
Looks good.

bangerth commented:
This is in reference to #314.
